File: app/src/main/java/com/averycorp/prismtask/domain/usecase/SelfCareNudgeEngine.kt
package com.averycorp.prismtask.domain.usecase

/**
 * A single nudge candidate produced by [SelfCareNudgeEngine] (v1.4.0 V2).
 *
 * The [id] is a stable key so the UI can persist "don't show this
 * consecutive time" state; the [kind] bucket lets the UI pick an icon.
 */
data class SelfCareNudge(val id: String, val kind: NudgeKind, val message: String)

enum class NudgeKind {
    REST_BREAK,
    BURNOUT_WARNING,
    MOVEMENT,
    WIND_DOWN
}

/**
 * Pure-function nudge selector (v1.4.0 V2).
 *
 * Consumes the burnout score, the current self-care ratio vs the target,
 * the time of day, and the last shown nudge id. Returns a nudge to show
 * right now, or `null` if nothing is warranted.
 *
 * The four nudge types rotate — if the same kind was just shown, the
 * engine picks a different one. This matches the vision deck's "rotate,
 * don't repeat the same one consecutively" requirement.
 */
class SelfCareNudgeEngine {
    fun select(
        burnoutScore: Int,
        selfCareRatio: Float,
        selfCareTarget: Float,
        hourOfDay: Int,
        lastShownId: String?
    ): SelfCareNudge? {
        val ratioBelowTarget = selfCareRatio < (selfCareTarget - LOW_SELF_CARE_BUFFER)
        val burnoutElevated = burnoutScore > BURNOUT_NUDGE_THRESHOLD
        if (!ratioBelowTarget && !burnoutElevated) return null

        val candidates = mutableListOf<SelfCareNudge>()

        // Always-eligible nudges when conditions hit.
        candidates.add(
            SelfCareNudge(
                id = "rest_break",
                kind = NudgeKind.REST_BREAK,
                message = "You haven't logged any self-care today — how about a 15-minute break?"
            )
        )
        if (burnoutElevated) {
            candidates.add(
                SelfCareNudge(
                    id = "burnout_warning",
                    kind = NudgeKind.BURNOUT_WARNING,
                    message = "Your burnout score is rising. Block 30 minutes for something you enjoy."
                )
            )
        }
        candidates.add(
            SelfCareNudge(
                id = "movement",
                kind = NudgeKind.MOVEMENT,
                message = "Movement reminder: a short walk can reset your focus."
            )
        )
        if (hourOfDay >= 18) {
            candidates.add(
                SelfCareNudge(
                    id = "wind_down",
                    kind = NudgeKind.WIND_DOWN,
                    message = "Wind-down suggestion: consider stopping work tasks for the evening."
                )
            )
        }

        // Rotate: skip any candidate with the same id as the last one shown.
        val filtered = candidates.filter { it.id != lastShownId }
        return (filtered.firstOrNull() ?: candidates.firstOrNull())
    }

    companion object {
        const val BURNOUT_NUDGE_THRESHOLD = 50
        const val LOW_SELF_CARE_BUFFER = 0.10f
    }
}

File: web/src/utils/selfCareNudgeEngine.ts
/**
 * Pure-function self-care nudge selector. Mirrors Android's
 * `SelfCareNudgeEngine.kt` (parity audit C.1e).
 *
 * Inputs:
 *  - `burnoutScore` (0..100; produced by `burnoutScorer.ts`)
 *  - `selfCareRatio` (0..1; from `BalanceTracker` currentRatios)
 *  - `selfCareTarget` (0..1; user-configured)
 *  - `hourOfDay` (0..23)
 *  - `lastShownId` (used to rotate away from the same kind back-to-back)
 *
 * Returns a nudge to show right now, or `null` if neither overload nor a
 * low-self-care signal warrants one.
 */

export type SelfCareNudgeKind =
  | 'rest_break'
  | 'burnout_warning'
  | 'movement'
  | 'wind_down';

export interface SelfCareNudge {
  id: string;
  kind: SelfCareNudgeKind;
  message: string;
}

const BURNOUT_NUDGE_THRESHOLD = 50;
const LOW_SELF_CARE_BUFFER = 0.1;

export function selectSelfCareNudge(opts: {
  burnoutScore: number;
  selfCareRatio: number;
  selfCareTarget: number;
  hourOfDay: number;
  lastShownId?: string | null;
}): SelfCareNudge | null {
  const { burnoutScore, selfCareRatio, selfCareTarget, hourOfDay, lastShownId } = opts;

  const ratioBelowTarget = selfCareRatio < selfCareTarget - LOW_SELF_CARE_BUFFER;
  const burnoutElevated = burnoutScore > BURNOUT_NUDGE_THRESHOLD;
  if (!ratioBelowTarget && !burnoutElevated) return null;

  const candidates: SelfCareNudge[] = [];

  candidates.push({
    id: 'rest_break',
    kind: 'rest_break',
    message:
      "You haven't logged any self-care today — how about a 15-minute break?",
  });
  if (burnoutElevated) {
    candidates.push({
      id: 'burnout_warning',
      kind: 'burnout_warning',
      message:
        'Your burnout score is rising. Block 30 minutes for something you enjoy.',
    });
  }
  candidates.push({
    id: 'movement',
    kind: 'movement',
    message: 'Movement reminder: a short walk can reset your focus.',
  });
  if (hourOfDay >= 18) {
    candidates.push({
      id: 'wind_down',
      kind: 'wind_down',
      message:
        'Wind-down suggestion: consider stopping work tasks for the evening.',
    });
  }

  // Rotate away from the same id back-to-back.
  const filtered = candidates.filter((c) => c.id !== lastShownId);
  return filtered[0] ?? candidates[0] ?? null;
}
